### PR TITLE
Determine which liquidSplitPct to use

### PR DESCRIPTION
**Pull request by misicnenad**
## Instructions on making this work

- run `yarn` or `yarn install` to install npm dependencies
- run `yarn run test --watchAll` to verify all tests still pass
- (optional) run `yarn run build` to verify the build passes
- run `yarn start` to start the webapp
-

## UI changes for review

When major UI changes are introduced with a PR, please include links to URLS to compare or screenshots demonstrating the difference and notify on design changes

### CODE REVIEW

File: src/components/donation/Steps/CurrentStep.tsx [hunk 1]
```diff
@@ -22,7 +25,7 @@ export default function CurrentStep({ config }: Props) {
     return (
       <Splits
         {...state}
-        liquidSplitPct={state.liquidSplitPct ?? config?.liquidSplitPct}
+        liquidSplitPct={liquidSplitPct}
```
Comment (ap-justin) [inline src/components/donation/Steps/CurrentStep.tsx:28]: I tried to update preview slider (`88`)  - but when I click `update snippet`. The previev slider didn't follow (back to `50`)

image

File: src/components/donation/Steps/useDetermineLiquidSplitPct.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import { useEffect, useRef, useState } from "react";
+import { DonationState, SplitsStep } from "slices/donation";
+import { Config } from "./types";
+
+/** Determines which liquidSplitPct value to use, depending on which was the most recently changed one */
+export default function useDetermineLiquidSplitPct(
+  config: Config | null,
+  state: DonationState
```
[the remark this answers is not in the document]
Comment (misicnenad) [reply src/components/donation/Steps/useDetermineLiquidSplitPct.ts:8]: My thinking was that `Splits` should only concern itself with displaying whatever value was passed to it and to extract the actual logic determining which value to use outside of it. That way we make `Splits` more independent of the context in which it's used.
If you don't think that's important ATM, the solution would basically become a different version of your PR https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/2817/files, so we'd close this (my) draft use that (your draft).

File: src/components/donation/Steps/CurrentStep.tsx (hunk 1 above)
Comment (misicnenad) [reply src/components/donation/Steps/CurrentStep.tsx:28]: This is a very weird bug. I updated [the useEffect responsible for updating the liquid %](https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/2819/files#diff-9c40ba9a8234eefaaca353ef2f5fb9fce9c968c7ba5459a1d78ac75780ac3235R41) and verified that the nested `setState` gets called:
```ts
  useEffect(() => {
    if (config?.liquidSplitPct) {
      setLiquidSplitPct(config.liquidSplitPct); // <-- this gets called!
    }
  }, [config, config?.liquidSplitPct]);
```
Despite this, the actual state value (`liquidSplitPct`) **does not get updated** :exploding_head:

File: src/components/donation/Steps/useDetermineLiquidSplitPct.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (misicnenad) [reply src/components/donation/Steps/useDetermineLiquidSplitPct.ts:8]: > This is a valid reason, the tradeoff is that CurrentStep has new detail other than choosing which step to render

Fair point.

> useDetermineSplit also has other responsibility (distinguishing a step from the other, which is a duplicate of what <CurrentStep/> is doing) other than detemrining split

That would unfortunately be the price to pay...

> If you tested my proposed solution to be working, I suggest we be practical and use it to finally close the widget update PR. I'll also consider it as a gesture of allowing me to help/contribute with the problem 🫡

Sounds good to me ser 🫡 Closing this PR.

One thing to note is that both of solutions feel... hacky. Even the "split redux actions" solution would involve some type of logic to determine when to update the state liquid % to the config's value. I guess it's the consequence of the complex requirement (make it all work on donate widget, widget config and regular donate flow while all 3 having slightly different needs).

**Comment by misicnenad** [comment]
Closing, see https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/2819#discussion_r1542758574

**Pull request closed without merging**